dali/test/python/test_external_source_parallel_custom_serialization.py

```diff
@@ -0,0 +1,432 @@
+
+
+def standard_global_callback(sample_info):
+    return np.full((10, 20), sample_info.idx_in_batch, dtype=np.uint8)
+
+
+def callback_idx(i):
+    return np.full((10, 20), i, dtype=np.uint8)
+
+
+@dali_pickle.pickle_by_value
+def callback_idx_by_value(i):
+    return np.full((10, 20), i, dtype=np.uint8)
+
+
+class NoDumpsParam(ValueError):
+    pass
+
+
+def dumps(obj, **kwargs):
+    if kwargs.get('special_dumps_param') != 42:
+        raise NoDumpsParam("Expected special_dumps_pram among kwargs, got {}".format(kwargs))
+    return dali_pickle.dumps(obj)
+
+
+def loads(data, **kwargs):
+    callbacks = dali_pickle.loads(data)
+    if kwargs.get('special_loads_param') == 84:
+        return [cb if cb.__name__ != 'callback_const_84' else callback_const_42 for cb in callbacks]
+    return callbacks
+
+
+# Register dummy reducer for custom type to check if DALI pickler did not interfere with
+# ability to register custom reducers for user defined types
+class DummyCb:
+
+    def __call__(self, sample_info):
+        return np.int32([1])
+
+class DummyCb42:
+
+    def __call__(self, sample_info):
+        return np.int32([42])
+
+
+def crazy_reducer(obj):
+    return DummyCb42().__reduce__()
+
+
+copyreg.pickle(DummyCb, crazy_reducer)
+
+
+# use process pid to assert that arrays were passed by value
+# and are equal for externalsource run in the same and separate process
+global_numpy_arrays = [np.full((10, 10), os.getpid() + i) for i in range(30)]
+
+
+def create_closure_callback_numpy(shape, data_set_size):
+    # use process pid to assert that arrays were passed by value
+    # and are equal for externalsource run in the same and separate process
+    data = [np.full(shape, os.getpid()) for _ in range(data_set_size)]
+
+    def callback(sample_info):
+        if sample_info.idx_in_epoch >= data_set_size:
+            raise StopIteration
+        return data[sample_info.idx_in_epoch]
+    
+    return callback
+
+
+def create_closure_callback_img_reader(data_set_size):
+    data_root = get_dali_extra_path()
+    images_dir = os.path.join(data_root, 'db', 'single', 'jpeg')
+
+    with open(os.path.join(images_dir, "image_list.txt"), 'r') as f:
+        file_label = [line.rstrip().split(' ') for line in f if line != '']
+        files, labels = zip(*file_label)
+
+    def py_file_reader(sample_info):
+        if sample_info.idx_in_epoch >= data_set_size:
+            raise StopIteration
+        sample_idx = sample_info.idx_in_epoch % len(files)
+        jpeg_filename = files[sample_idx]
+        label = np.int32([labels[sample_idx]])
+        with open(os.path.join(images_dir, jpeg_filename), 'rb') as f:
+            encoded_img = np.frombuffer(f.read(), dtype=np.uint8)
+        return encoded_img, label
+
+    return py_file_reader
+
+
+def create_simple_pipeline(callback, py_callback_pickler, batch_size, parallel=True, py_num_workers=None):
+
+    extra = {}
+    if py_callback_pickler != UseDefault:
```

Comment:
Why not:
```suggestion
    if py_callback_pickler is nt None:
```
like we do elsewhere?

Reply:
done